Update backend.Key to be its own distinct type

rosstimothy opened the pull request:
Converts the `backend.Key` from a slice of bytes to a concrete struct. The main motivation behind this change is to be able to better distinguish individual components of a key. The textual representation of a backend key is constructed by joining all components of the key with a `/`. By representing the entire key as a single textual object it prevented resources containing a `/` in their name from being properly identified. There was a lot of code that interpolated keys in various manners and expected only a specific number of subcomponents for a particular resource. This lead to most, if not all, of the RPCs used to create a resource to permit a name containing `/`, (i.e. a user named `test/llama/1`), but any RPCs used to retrieve the resources would fail to retrieve them from the backend. 

Every `backend.Key` now carries a slice of all the individual components in addition to the textual representation. This permits more fine grained inspection per component, while also not having to construct the textual representation for a group of components more than once.

The `backend.Sanitizer` was updated to only validate keys for backend writes. Retrieving and deleting invalid of malformed keys is however permitted. This will allow any existing resources that were created with `/` in one of the components to become visible in the UI or tctl get and deletable. However, any new resources with ` /` in there name are explicitly prevented.

Closes #6088 
Closes #9107
Closes #10576
Closes #42823

espadolini commented on lib/backend/key.go:
```diff
@@ -19,81 +19,138 @@ package backend
-	return internalKey("", parts...)
+func NewKey(components ...string) Key {
+	k := internalKey("", components...)
+	k.exactKey = k.s == SeparatorString || (len(k.s) > 0 && k.s[len(k.s)-1] == Separator)
```
```suggestion
	k.exactKey = len(k.s) > 0 && k.s[len(k.s)-1] == Separator
```

Isn't `internalKey` always returning a nonempty key, anyway?

fspmarshall commented on lib/backend/key.go:
```diff
@@ -19,81 +19,138 @@ package backend
+	return len(k.components) == 0 && k.s == ""
+}
+
+func internalKey(internalPrefix string, components ...string) Key {
```
+1, it seems like that should be the default behavior, and silently splitting components should be some kind of opt-in for back-combat.

fspmarshall commented on lib/backend/key.go:
```diff
@@ -19,81 +19,138 @@ package backend
+	return len(k.components) == 0 && k.s == ""
+}
+
+func internalKey(internalPrefix string, components ...string) Key {
```
This function is pretty dense.  I think some detailed comments would be helpful.

fspmarshall reviewed (approved):
LGTM once existing feedback is addressed.

espadolini commented on lib/services/local/trust_test.go:
```diff
@@ -132,7 +132,7 @@ func TestRemoteClusterCRUD(t *testing.T) {
 	err = trustService.DeleteRemoteCluster(ctx, "foo")
 	require.Error(t, err)
-	require.ErrorIs(t, err, trace.NotFound("key \"/remoteClusters/foo\" is not found"))
+	require.ErrorIs(t, err, trace.NotFound("key /remoteClusters/foo is not found"))
```
It's probably fine now that we're better at enforcing key sanitization? 🤷

rosstimothy committed:
Update backend.Key to be its own distinct type

Converts the backend.Key from a slice of bytes to a concrete struct.
The main motivation behind this change is to be able to better
distinguish individual components of a key. The textual representation
of a backend key is constructed by joining all components of the key
with a /. By representing the entire key as a single textual object
it prevented resources containing a / in their name from being
properly identified. There was a lot of code that interpolated keys
in various manners and expected only a specific number of
subcomponents for a particular resource. This lead to most,
if not all, of the RPCs used to create a resource to permit a name
containing /, (i.e. a user named test/llama/1), but any RPCs used
to retrieve the resources would fail to retrieve them from the
backend.

Every backend.Key now carries a slice of all the individual
components in addition to the textual representation. This permits
more fine grained inspection per component, while also not having
to construct the textual representation for a group of components
more than once.

The backend.Sanitizer was updated to only validate keys for backend
writes. Retrieving and deleting invalid of malformed keys is now
permitted. This will allow any existing resources that were created
with / in one of the components to become retrievable and deletable.
However, any new resources with  / in their name are explicitly
prevented.

Closes #6088
Closes #9107
Closes #10576
Closes #42823
